### ISSUE TITLE
[CommonHTML] optional xhtml compatibility [was: img tags not closed]

### ISSUE THREAD

**Issue by pkra** [mathjax/MathJax-node#228]
When using `mglyph`'s, the result has unclosed (not self-closing) img tags.

E.g.,

```html
<math xmlns="http://www.w3.org/1998/Math/MathML" display="block">
  <mi><mglyph width="1" height="1" src="bla.png" alt="Bla"/></mi>
</math>
```

produces 

```html
<span class="mjx-chtml MJXc-display" style="text-align: center;"><span id="MathJax-Element-1-Frame" class="mjx-chtml"><span id="MJXc-Node-1" class="mjx-math" role="math"><span id="MJXc-Node-2" class="mjx-mrow"><span id="MJXc-Node-3" class="mjx-mi"><span id="MJXc-Node-4" class="mjx-mglyph"><img src="bla.png" alt="Bla" title="Bla" style="width: 0px; height: 0px; vertical-align: 0;"></span></span></span></span></span></span>
```

**Comment by dpvc** [comment]
This is correct HTML output.  The `<img>` tag does not need a close tag in HTML (this is not XML).  In any case, it is jsdom that is producing the output, here, so we don't control that directly.  One would have to post-process the result in order to fix it.

**Comment by pkra** [comment]
Sorry I obviously wasn't thinking straight when posting. This came out of an XML workflow and I was in a hurry.

So let me rephrase this to a feature request for xhtml compatibility. This would be particularly useful for ebook production (though I ran into it in a more general XML context).

jsdom has an option for that but it has limitations, https://github.com/tmpvar/jsdom/issues/885; I'll try it out if there's interest in such a feature.

**Comment by pkra** [comment]
In the mean time, `data.html.replace(/(<img[^>]*)/,"$1 \/");` might help? (I'm no regex magician.)

**Comment by pkra** [comment]
FWIW, I couldn't figure out the jsdom side of this. Might be the serialization problem mentioned at https://github.com/tmpvar/jsdom/issues/798. 

I'm not sure this is worth pursuing.

**Comment by dpvc** [comment]
> Might be the serialization problem mentioned at tmpvar/jsdom#798.

Yes, it is certainly a serialization problem, not a parsing one, and they would need different serialization rules for XHTML and HTML files, which I don't think they have.

Note that for HTML files (as opposed to XHTML ones), the browsers **don't** retain the redundant closing slash.  E.g., try the file

```
<!DOCTYPE html>
<html>
<head>
<title>Test of serialization of self-closing tags in HTML</title>
<script>
window.onload = function () {alert(document.body.innerHTML)}
</script>
</head>
<body>
<img src="test.png"/><br/>
</body>
</html>
```
and see.

**Comment by dpvc** [comment]
> In the mean time, `data.html.replace(/(<img[^>]*)/g,"$1 \/");` might help?

It would have to be `data.html = data.html.replace(...)`.  The problem would be if there was anything that contained `<img` that wasn't actually an image.  Unfortunately, it is possible to arrange for that.  For example,

    bin/mml2html '<math><mglyph src="tmp.png" alt="&lt;img&gt;"></mglyph></math>'

produces the output

    <span class="mjx-chtml"><span class="mjx-math" role="math"><span class="mjx-mrow"><span class="mjx-mglyph"><img src="tmp.png" alt="<img>" title="<img>" style="vertical-align: 0;"></span></span></span></span>

and so the regular expression would match `<img src="tmp.png" alt="<img` rather than the full image tag, and then would match the second `<img` in the `title` as well.  So this would insert the `/` in two wrong places and miss the correct place all together.

Similarly, the command

    bin/mml2html '<math><mi id="imf&lt;img">x</mi></math>'

produces the output

    <span class="mjx-chtml"><span class="mjx-math" role="math"><span class="mjx-mrow"><span id="imf<img" class="mjx-mi"><span class="mjx-char MJXc-TeX-math-I" style="padding-top: 0.225em; padding-bottom: 0.298em;">x</span></span></span></span></span>

and the regular expression will match the `<img" class="mjx-mi"` that comes from the `id` in the `mjx-mi` span, and so incorrectly add a `/` to the end of that span.

It is because of these kinds of situations that manipulating HTML via regular expressions is dangerous.

Technically, jsdom should be converting the HTML special characters in attributes into entities, but it is not strictly necessary, as the browsers parse the jsdom output correctly.  It's just bad practice.  Had they done that, your regular expression would work, but as it is, it can cause problems.

I seem to recall having had to do my own serialization for the `toMathML()` function, at least in some browsers (probably IE) since they didn't convert special characters.  I suppose we could introduce our own serializer rather than using `innerHTML`.  It could handle the conversion of special characters, the removal of automatically created id's, and the addition of self-closing slashes (optionally).  This would avoid these regular expression pitfalls, but it would introduce overhead.

**Comment by pkra** [comment]
> It is because of these kinds of situations that manipulating HTML via regular expressions is dangerous.

Indeed. 

> I suppose we could introduce our own serializer rather than using innerHTML.

Sounds like 3.0 territory?

**Comment by dpvc** [comment]
> Sounds like 3.0 territory?

Fine by me.

**Comment by pkra** [comment]
Well, I just learned that e.g libxml can actually parse an HTML string just fine. Perhaps most do, in which case this should really just be a downstream concern. So I'm ok with just closing this (in which case: sorry for the noise).

**Comment by dpvc** [comment]
I'm OK with closing, but we might want to keep XHTML in mind for 3.0.

**Comment by pkra** [comment]
Note from the future:  jsdom/jsdom#798 was fixed; it is now possible to configure jsdom to produce xhtml output.
